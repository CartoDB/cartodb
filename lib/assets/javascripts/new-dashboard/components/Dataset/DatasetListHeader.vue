<template>
  <div class="dataset-list-row">
    <div class="dataset-list-cell cell--start">
    </div>
    <div class="dataset-list-cell cell--main" @click="changeOrder('name')">
      <span class="text element-sort is-small is-txtSoftGrey"
            :class="{ 'is-active': isOrderApplied('name'), 'is-reversed': isReverseOrderApplied('name') }">
        {{ $t(`DatasetListHeader.name`) }}
      </span>
    </div>
    <div class="dataset-list-cell cell--large" @click="changeOrder('updated_at')">
      <span class="text element-sort is-small is-txtSoftGrey"
            :class="{ 'is-active': isOrderApplied('updated_at'), 'is-reversed': isReverseOrderApplied('updated_at') }">
        {{ $t(`DatasetListHeader.lastModified`) }}
      </span>
    </div>
    <div class="dataset-list-cell cell--small" @click="changeOrder('estimated_row_count')">
      <span class="text element-sort is-small is-txtSoftGrey"
            :class="{ 'is-active': isOrderApplied('estimated_row_count'), 'is-reversed': isReverseOrderApplied('estimated_row_count') }">
        {{ $t(`DatasetListHeader.rows`) }}
      </span>
    </div>
    <div class="dataset-list-cell cell--small" @click="changeOrder('size')">
      <span class="text element-sort is-small is-txtSoftGrey"
            :class="{ 'is-active': isOrderApplied('size'), 'is-reversed': isReverseOrderApplied('size') }">
        {{ $t(`DatasetListHeader.size`) }}
      </span>
    </div>
    <div class="dataset-list-cell cell--small" @click="changeOrder('dependent_visualizations')">
      <span class="text element-sort is-small is-txtSoftGrey"
            :class="{ 'is-active': isOrderApplied('dependent_visualizations'), 'is-reversed': isReverseOrderApplied('dependent_visualizations') }">
        {{ $t(`DatasetListHeader.usage`) }}
      </span>
    </div>
    <div class="dataset-list-cell cell--small cell--privacy" @click="changeOrder('privacy')">
      <span class="text element-sort is-small is-txtSoftGrey"
            :class="{ 'is-active': isOrderApplied('privacy'), 'is-reversed': isReverseOrderApplied('privacy') }">
        {{ $t(`DatasetListHeader.privacy`) }}
      </span>
    </div>
    <div class="dataset-list-cell cell--end"></div>
  </div>
</template>

<script>
export default {
  name: 'DatasetListHeader.',
  props: {
    order: String,
    orderDirection: String
  },
  methods: {
    changeOrder (order) {
      if (this.order === order) {
        return this.setOrder(
          order,
          this.getOppositeOrderDirection(this.orderDirection)
        );
      }

      this.setOrder(order);
    },

    getOppositeOrderDirection (order) {
      if (order === 'desc') {
        return 'asc';
      }

      if (order === 'asc') {
        return 'desc';
      }
    },

    isOrderApplied (order) {
      return order === this.order;
    },

    isReverseOrderApplied (order) {
      return order === this.order && this.orderDirection === 'asc';
    },

    setOrder (order, direction = 'desc') {
      this.$emit('changeOrder', { order, direction });
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.dataset-list-row {
  display: flex;
  align-items: center;
  width: 100%;
  margin-bottom: 0;
  padding: 8px 14px;
  overflow: hidden;
  background-color: $white;
}

.dataset-list-cell {
  position: relative;
  flex-grow: 0;
  flex-shrink: 0;
  padding: 0 10px;
  text-overflow: ellipsis;
  white-space: nowrap;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }
}

.cell--start {
  display: flex;
  align-items: center;
  align-self: flex-start;
  width: 46px;
  height: 100%;
}

.cell--end {
  width: 34px;
}

.cell--main {
  flex-grow: 1;
  flex-shrink: 1;
  min-width: 200px;
}

.cell--large {
  width: 165px;
}

.cell--medium {
  width: 120px;
}

.cell--small {
  width: 80px;
}

.cell--privacy {
  display: flex;
  align-items: center;
}

.element-sort {
  &.is-active,
  &:hover {
    color: $text-color;
    cursor: pointer;

    &::after {
      content: '';
      position: absolute;
      width: 14px;
      height: 100%;
      margin-left: 4px;
      transition: all 0.25s cubic-bezier(0.4, 0.01, 0.165, 0.99);
      background-image: url('../../assets/icons/datasets/chevron.svg');
      background-repeat: no-repeat;
      background-position: center;
    }

    &.is-reversed {
      &::after {
        transform: rotate(180deg);
      }
    }
  }
}
</style>
