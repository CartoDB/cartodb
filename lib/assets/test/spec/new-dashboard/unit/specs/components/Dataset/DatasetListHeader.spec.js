import { shallowMount } from '@vue/test-utils';
import DatasetListHeader from 'new-dashboard/components/Dataset/DatasetListHeader';

const $t = key => key;

describe('DatasetListHeader.vue', () => {
  let datasetListHeader;
  beforeEach(() => {
    datasetListHeader = createDatasetListHeader();
  });

  it('should render correct contents', () => {
    expect(datasetListHeader).toMatchSnapshot();
  });

  describe('Methods', () => {
    describe('changeOrder', () => {
      it('should call setOrder with passed order', () => {
        const setOrder = jest.fn();

        const header = createDatasetListHeader({
          methods: { setOrder }
        });

        header.vm.changeOrder('updated_at');

        expect(setOrder).toHaveBeenCalledWith('updated_at');
      });

      it('should call setOrder with opposite order if passed order was applied already', () => {
        const setOrder = jest.fn();

        const header = createDatasetListHeader({
          methods: { setOrder }
        });

        header.vm.changeOrder('size');

        expect(setOrder).toHaveBeenCalledWith('size', 'asc');
      });
    });

    describe('getOppositeOrderDirection', () => {
      it('should return asc when passing desc', () => {
        expect(datasetListHeader.vm.getOppositeOrderDirection('asc')).toBe('desc');
      });

      it('should return desc when passing asc', () => {
        expect(datasetListHeader.vm.getOppositeOrderDirection('desc')).toBe('asc');
      });
    });

    describe('isOrderApplied', () => {
      it('should return true if order is applied', () => {
        expect(datasetListHeader.vm.isOrderApplied('size')).toBe(true);
      });
    });

    describe('isReversedOrderApplied', () => {
      it('should return true if reverse order is applied', () => {
        datasetListHeader.setData({
          order: 'size',
          orderDirection: 'asc'
        });
        expect(datasetListHeader.vm.isReverseOrderApplied('size')).toBe(true);
      });
    });

    describe('setOrder', () => {
      it("should emit 'changeOrder' event with order and direction", () => {
        datasetListHeader.vm.setOrder('updated_at', 'asc');
        expect(datasetListHeader.emitted('changeOrder')[0]).toEqual([
          { direction: 'asc', order: 'updated_at' }
        ]);
      });
    });
  });
});

function createDatasetListHeader (overridenComponentOptions) {
  return shallowMount(DatasetListHeader, {
    mocks: {
      $t
    },
    propsData: {
      order: 'size',
      orderDirection: 'desc'
    },
    ...overridenComponentOptions
  });
}
