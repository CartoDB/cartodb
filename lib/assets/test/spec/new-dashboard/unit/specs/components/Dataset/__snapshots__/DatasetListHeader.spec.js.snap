// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetListHeader.vue should render correct contents 1`] = `
<div class="dataset-list-row">
  <div class="dataset-list-cell cell--start"></div>
  <div class="dataset-list-cell cell--main"><span class="text element-sort is-small is-txtSoftGrey">
      DatasetListHeader.name
    </span></div>
  <div class="dataset-list-cell cell--large"><span class="text element-sort is-small is-txtSoftGrey">
      DatasetListHeader.lastModified
    </span></div>
  <div class="dataset-list-cell cell--small"><span class="text element-sort is-small is-txtSoftGrey">
      DatasetListHeader.rows
    </span></div>
  <div class="dataset-list-cell cell--small"><span class="text element-sort is-small is-txtSoftGrey is-active">
      DatasetListHeader.size
    </span></div>
  <div class="dataset-list-cell cell--small"><span class="text element-sort is-small is-txtSoftGrey">
      DatasetListHeader.usage
    </span></div>
  <div class="dataset-list-cell cell--small cell--privacy"><span class="text element-sort is-small is-txtSoftGrey">
      DatasetListHeader.privacy
    </span></div>
  <div class="dataset-list-cell cell--end"></div>
</div>
`;
